op-node/rollup/derive/blob_data_source.go

```diff
@@ -0,0 +1,172 @@
+// transactions are found. Returns nil array whenever an error is returned.
+func (ds *BlobDataSource) open(ctx context.Context) ([]blobOrCalldata, error) {
+	_, txs, err := ds.fetcher.InfoAndTxsByHash(ctx, ds.ref.Hash)
+	if err != nil {
+		if errors.Is(err, ethereum.NotFound) {
+			return nil, NewResetError(fmt.Errorf("%w: failed to open blob data source", err))
+		}
+		return nil, NewTemporaryError(fmt.Errorf("%w: failed to open blob data source", err))
+	}
+
+	data, hashes := dataAndHashesFromTxs(txs, &ds.dsCfg, ds.batcherAddr)
+
+	if len(hashes) == 0 {
+		// there are no blobs to fetch so we can return immediately
+		return data, nil
+	}
+
+	// download the actual blob bodies corresponding to the indexed blob hashes
+	blobs, err := ds.blobsFetcher.GetBlobs(ctx, ds.ref, hashes)
+	if errors.Is(err, ethereum.NotFound) {
+		// If the L1 block was available, then the blobs should be available too. The only
+		// exception is if the blob retention window has expired, which we will ultimately handle
+		// by failing over to a blob archival service.
+		return nil, NewResetError(fmt.Errorf("%w: failed to fetch blobs", err))
+	} else if err != nil {
+		return nil, NewTemporaryError(fmt.Errorf("%w: failed to fetch blobs", err))
+	}
+
+	// go back over the data array and populate the blob pointers
+	if err := fillBlobPointers(data, blobs); err != nil {
+		// this shouldn't happen unless there is a bug in the blobs fetcher
+		return nil, NewResetError(fmt.Errorf("%w: failed to fill blob pointers", err))
+	}
+	return data, nil
+}
+
+// dataAndHashesFromTxs extracts calldata and datahashes from the input transactions and returns them. It
+// creates a placeholder blobOrCalldata element for each returned blob hash that must be populated
+// by fillBlobPointers after blob bodies are retrieved.
+func dataAndHashesFromTxs(txs types.Transactions, config *DataSourceConfig, batcherAddr common.Address) ([]blobOrCalldata, []eth.IndexedBlobHash) {
+	data := []blobOrCalldata{}
+	var hashes []eth.IndexedBlobHash
+	blobIndex := 0 // index of each blob in the block's blob sidecar
+	for _, tx := range txs {
+		// skip any non-batcher transactions
+		if !isValidBatchTx(tx, config.l1Signer, config.batchInboxAddress, batcherAddr) {
+			blobIndex += len(tx.BlobHashes())
+			continue
+		}
+		// handle non-blob batcher transactions by extracting their calldata
+		if tx.Type() != types.BlobTxType {
+			calldata := eth.Data(tx.Data())
+			data = append(data, blobOrCalldata{nil, &calldata})
+			continue
+		}
+		// handle blob batcher transactions by extracting their blob hashes, ignoring any calldata.
+		if len(tx.Data()) > 0 {
+			log.Warn("blob tx has calldata, which will be ignored", "txhash", tx.Hash())
+		}
+		for _, h := range tx.BlobHashes() {
+			idh := eth.IndexedBlobHash{
+				Index: uint64(blobIndex),
+				Hash:  h,
+			}
+			hashes = append(hashes, idh)
+			data = append(data, blobOrCalldata{nil, nil}) // will fill in blob pointers after we download them below
```

Comment:
Abstraction feels a bit weird here that its a `nil, nil` used and then filled in outside of the function

Reply:
agree it's unusual; conceptually it's just a part of the open() call, but I wanted to be able to unit test it separately, and a fancier abstraction didn't seem to be worth it. Open to ideas here though.

Reply:
I agree slightly weird, but after reading the flow, it does make some sense to split up finding the blobs from fetching the blobs. I think if we wanted to clean it up, we'd want to make blobOrCalldata be something like the following rust enum or maybe have an internal enum then transform it to an external enum.

```
enum blobOrCalldata {
    Calldata(Vec<u8>),
    UnInitBlob,
    InitBlob(...),
}
```

Reply:
I experimented with replacing the union type with an interface [here](https://github.com/ethereum-optimism/optimism/commit/f1b00980064b3b794df092957eb6104ac12bb77e). Not sure if it actually improves things or makes them less readable. It basically removes 1 or 2 `if`s by introducing abstraction.
It feels like it's only half-way there. A proper way would probably be to create a struct around the `[]dataItem` that counts the blob items as they are created and filled.
Feel free to just ignore it ;)

Reply:
Yeah agree it only feels half-way there.  Let's revisit later?

Reply:
yep good idea